### PR TITLE
:seedling:Upgrade golang version (1.19.6 -> 1.20.3)

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
**What this PR does / why we need it**:
- Upgrade golang version (1.19.6 -> 1.20.3)
- Fixing golangci-lint error: rand.Seed has been deprecated since Go 1.20.
Seeding is added by default from 1.20 onwards. More info [here](https://github.com/golang/go/blob/3f8f929d60a90c4e4e2b07c8d1972166c1a783b1/src/math/rand/rand.go#L383)

Part of https://github.com/kubernetes-sigs/cluster-api/issues/8459

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
/test pull-cluster-api-apidiff-main

This looks good to me, thanks!
@sbueringer do we take care of the [go mod](https://github.com/kubernetes-sigs/cluster-api/blob/main/go.mod#L3) bump in a separate PR (as we usual did in the past)?

**Comment by furkatgofurov7** [comment]
> test pull-cluster-api-apidiff-main

hmm, seems I can't retrigger the test, might be that only maintainers have the rights to do so, but let's try these too since it was triggered but there was a force push after

/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-27-latest-main

**Comment by sbueringer** [comment]
> hmm, seems I can't retrigger the test, might be that only maintainers have the rights to do so, but let's try these too since it was triggered but there was a force push after

There's no limitations regarding triggering jobs (at least if you're an org member)

**Review by killianmuldoon — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
@Prajyot-Parab please add "Part of #8459" to the PR description

File: .golangci.yml
Comment (sbueringer) [inline .golangci.yml:2]: Let's please also bump:
* go.mod files

In a follow-up:
* cloudbuild.yaml, cloudbuild-nightly.yaml: there is currently no gcb-docker-gcloud image with Go 1.20 => I think that was in test-infra and we can open a PR to bump Go in that image

**Review by sbueringer — CHANGES_REQUESTED** [review]

**Comment by sbueringer** [comment]
> @sbueringer do we take care of the [go mod](https://github.com/kubernetes-sigs/cluster-api/blob/main/go.mod#L3) bump in a separate PR (as we usual did in the past)?

I would just do it in this PR. Just a bit simpler. I think there is no need to split it into a separate PR.

**Comment by furkatgofurov7** [comment]
> > @sbueringer do we take care of the [go mod](https://github.com/kubernetes-sigs/cluster-api/blob/main/go.mod#L3) bump in a separate PR (as we usual did in the past)?
> 
> I would just do it in this PR. Just a bit simpler. I think there is no need to split it into a separate PR.

That is what I always thought but was not aware the reasoning behind the splitting, thanks for confirming. 

/cc @Prajyot-Parab (to confirm the answer ^ to your question offline)

File: .golangci.yml
Comment (Prajyot-Parab) [reply .golangci.yml:2]: Noted.

**Review by sbueringer — APPROVED** [review]

**Comment by sbueringer** [comment]
/lgtm
/approve

/hold
Feel free to remove the hold once all currently running tests are green

**Comment by sbueringer** [comment]
/retest

known flake

**Comment by sbueringer** [comment]
/hold cancel

tests are fine

**Comment by furkatgofurov7** [comment]
> @Prajyot-Parab Can you check if there's anything you have to do / can do regarding the CLA?

@sbueringer fixed

**Pull request merged**
